don't sync daemonsets with selectors that match all pods

mikedanese opened the pull request:
Fixes #23218

mikedanese committed:
don't sync daemonsets or controllers with selectors that match all pods

davidopp commented:
Can you add a test?

mikedanese commented:
Yup I'll add a test then reassign to you.

mikedanese commented:
@davidopp I've added a test.

davidopp commented:
Can you change the test so that the node has a label, and you create in the same test a second DaemonSet, which matches the node's labels, and verify that it schedules? (so that the only difference between the two daemon sets is that one has empty selector and one doesn't?) I realize there are lots of other tests in the file, but this makes the test more self-contained. Also can you change the name of the test so that it says WithEmptyLabelSelector (instead of WithoutLabelSelector) and add comment describing what the test does. Lastly, I noticed several tests in this file all have exactly the same comment describing them -- please fix (even though it's unrelated to this PR).

mikedanese commented:
@davidopp The node selector is ok. The selector is the one that is getting emptied out. As this test is currently written:

1. Schedule a pod on a node with a realistic label and not controlled by a daemonset
2. Create a daemonset that matches no nodes and matches all pods.

Before this change, this would cause the daemonset to try to delete all pods in the cluster because it would see them as miss-scheduled. After this change, we assert that the pod created in step 1. is not deleted. Do you just want me to test other scenarios?

mikedanese commented:
@davidopp PTAL, I added the scenario you were suggesting

davidopp commented:
LGTM

bgrant0607 commented:
Deployments have the same issue. I don't think it makes sense to fix one but not the other.

cc @janetkuo

mikedanese commented:
@bgrant0607 I mentioned that I didn't think deployments had the issue https://github.com/kubernetes/kubernetes/issues/23218#issuecomment-199493251 or that it wasn't exactly analagous. I had added the same feature in deployment controller and removed it. Do you want me to readd it?

https://github.com/kubernetes/kubernetes/issues/23218#issuecomment-199493251

bgrant0607 commented:
@mikedanese I responded on that issue. Yes, I think we need an analogous fix for Deployment.

mikedanese commented:
Ok, I will readd

bgrant0607 commented on pkg/controller/daemon/controller.go:
```diff
@@ -623,6 +625,12 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 
+	everything := unversioned.LabelSelector{}
+	if reflect.DeepEqual(ds.Spec.Selector, &everything) {
+		dsc.eventRecorder.Eventf(ds, api.EventTypeWarning, "SelectingAll", "This controller is selecting all pods. Skipping sync.")
```
As I mentioned on the Deployment PR, I think we need to improve this message.

mikedanese replied:
Ok, I will fix both
